src/dict.h

```diff
@@ -78,11 +70,23 @@ typedef struct dictType {
      * flag is set. */
 } dictType;
 
+typedef struct dictTypeExt {
+        /* Invoked at the start of dict initialization/rehashing (old and new ht are already created) */
+    void (*rehashingStarted)(dict *d);
+    /* Invoked at the end of dict initialization/rehashing of all the entries from old to new ht. Both ht still exists
+     * and are cleaned up after this callback.  */
+    void (*rehashingCompleted)(dict *d);
+    /* Allow a dict to carry extra caller-defined metadata. The
+     * extra memory is initialized to 0 when a dict is allocated. */
+    size_t (*dictMetadataBytes)(dict *d);
+} dictTypeExt;
+
 #define DICTHT_SIZE(exp) ((exp) == -1 ? 0 : (unsigned long)1<<(exp))
 #define DICTHT_SIZE_MASK(exp) ((exp) == -1 ? 0 : (DICTHT_SIZE(exp))-1)
 
 struct dict {
     dictType *type;
+    dictTypeExt *typeext;
```

Comment:
i rather avoid adding members to the `dict` struct.
maybe this can still be part of `dictType`?